firedrake/function.py

```diff
@@ -504,9 +504,12 @@ def at(self, arg, *args, **kwargs):
         if not arg.shape:
             arg = arg.reshape(-1)
 
+        mesh = self.function_space().mesh()
+        if mesh.variable_layers:
+            raise NotImplementedError("Point evaluation not implemented for variable layers")
         # Immersed not supported
-        tdim = self.function_space().mesh().ufl_cell().topological_dimension()
-        gdim = self.function_space().mesh().ufl_cell().geometric_dimension()
+        tdim = mesh.ufl_cell().topological_dimension()
+        gdim = mesh.ufl_cell().geometric_dimension()
```

Comment:
I think UFL `Mesh` has `.topological_dimension()` and `.geometric_dimension()` directly.